tests/pubsub/testgossipsub.nim

```diff
@@ -1237,3 +1301,434 @@ suite "GossipSub":
+    await sleepAsync(DURATION_TIMEOUT)
+
+    # At least one of the nodes should have received an iHave message
+    # The check is made this way because the mesh structure changes from run to run
+    check:
+      anyIt(receivedIHaves, it > 0)
+
+    await allFuturesThrowing(nodes.mapIt(allFutures(it.switch.stop())))
+
+  asyncTest "messages are not sent back to source or forwarding peer":
+    # Instantiate 3 nodes
+    let
+      numberOfNodes = 3
+      topic = "foobar"
+      nodes = generateNodes(numberOfNodes, gossip = true)
+      nodesFut = await allFinished(nodes.mapIt(it.switch.start()))
+      node0 = nodes[0]
+      node1 = nodes[1]
+      node2 = nodes[2]
+
+    # Each node with a handler
+    var
+      handlerFuture0 = newFuture[bool]()
+      handlerFuture1 = newFuture[bool]()
+      handlerFuture2 = newFuture[bool]()
+
+    proc handler0(topic: string, data: seq[byte]) {.async.} =
+      handlerFuture0.complete(true)
+
+    proc handler1(topic: string, data: seq[byte]) {.async.} =
+      handlerFuture1.complete(true)
+
+    proc handler2(topic: string, data: seq[byte]) {.async.} =
+      handlerFuture2.complete(true)
+
+    # Connect them in a ring
+    await node0.switch.connect(node1.peerInfo.peerId, node1.peerInfo.addrs)
+    await node1.switch.connect(node2.peerInfo.peerId, node2.peerInfo.addrs)
+    await node2.switch.connect(node0.peerInfo.peerId, node0.peerInfo.addrs)
+    await sleepAsync(DURATION_TIMEOUT)
+
+    # Subscribe them all to the same topic
+    nodes[0].subscribe(topic, handler0)
+    nodes[1].subscribe(topic, handler1)
+    nodes[2].subscribe(topic, handler2)
+    await sleepAsync(DURATION_TIMEOUT)
+
+    # When node 0 sends a message
+    discard nodes[0].publish(topic, "Hello!".toBytes())
+    await sleepAsync(DURATION_TIMEOUT)
+
+    # Nodes 1 and 2 should receive the message, but node 0 shouldn't receive it back
+    check:
+      (await handlerFuture0.waitForResult()).isErr
+      (await handlerFuture1.waitForResult()).isOk
+      (await handlerFuture2.waitForResult()).isOk
+
+    await allFuturesThrowing(nodes.mapIt(allFutures(it.switch.stop())))
+
+  asyncTest "flood publish to all peers with score above threshold, regardless of subscription":
+    # Given 3 nodes
+    let
+      numberOfNodes = 3
+      topic = "foobar"
+      nodes = generateNodes(numberOfNodes, gossip = true, floodPublish = true)
+      nodesFut = nodes.mapIt(it.switch.start())
+      g0 = GossipSub(nodes[0])
+
+    # Nodes 1 and 2 are connected to node 0
+    await nodes[0].switch.connect(nodes[1].peerInfo.peerId, nodes[1].peerInfo.addrs)
+    await nodes[0].switch.connect(nodes[2].peerInfo.peerId, nodes[2].peerInfo.addrs)
+
+    # Given 2 handlers
+    var
+      handlerFut1 = newFuture[bool]()
+      handlerFut2 = newFuture[bool]()
+
+    proc handler1(topic: string, data: seq[byte]) {.async.} =
+      handlerFut1.complete(true)
+
+    proc handler2(topic: string, data: seq[byte]) {.async.} =
+      handlerFut2.complete(true)
```

Comment:
this could be easily created with utility:

```
let handleFut1, handler1 = createCompleteHandler()
let handleFut2, handler2 = createCompleteHandler()

...

proc createCompleteHandler() ... return signature... =
    var fut = newFuture[bool]()
    proc handler(topic: string, data: seq[byte]) {.async.} =
      fut.complete(true)
   
   return fut, handler
```